data_lang/pretty.py

```diff
@@ -1,40 +1,546 @@
+        """Print one of two options (using '[', '_', ']' for open, sep, close):
+    
+        ```
+        [_mdoc]
+        ------
+        [
+            mdoc
+        ]
+        ```
+        """
+        return _Group(
+            _Concat([
+                _Text(open),
+                _Indent(self.indent, _Concat([_Break(sep), mdoc])),
+                _Break(""),
+                _Text(close)
+            ]))
+
+    def _Join(self, items, sep, space):
+        # type: (List[MeasuredDoc], str, str) -> MeasuredDoc
+        """Join `items`, using either 'sep+space' or 'sep+newline' between them."""
+        seq = [items[0]]
+        for item in items[1:]:
+            seq.append(_Text(sep))
+            seq.append(_Break(space))
+            seq.append(item)
+        return _Concat(seq)
+
+    def _DictKey(self, s):
+        # type: (str) -> MeasuredDoc
+        if match.IsValidVarName(s):
+            return self._Styled(KEY_STYLE, _Text(s))
         else:
-            # TODO: handle more cases
-            return doc.Newline
+            return self._Styled(KEY_STYLE,
+                                _Text(fastfunc.J8EncodeString(
+                                    s, True)))  # lossy_json=True
+
+    def _StringLiteral(self, s):
+        # type: (str) -> MeasuredDoc
+        return _Text(fastfunc.J8EncodeString(s, True))  # lossy_json=True
+
+    def _YshList(self, vlist):
+        # type: (value.List) -> MeasuredDoc
+        """Print a string literal."""
+        if len(vlist.items) == 0:
+            return _Text("[]")
+        mdocs = [self._Value(item) for item in vlist.items]
+        return self._Surrounded("[", "", self._Join(mdocs, ",", " "), "]")
+
+    def _YshDict(self, vdict):
+        # type: (value.Dict) -> MeasuredDoc
+        if len(vdict.d) == 0:
+            return _Text("{}")
+        mdocs = []
+        for k, v in iteritems(vdict.d):
+            mdocs.append(
+                _Concat([self._DictKey(k),
+                         _Text(": "),
+                         self._Value(v)]))
+        return self._Surrounded("{", "", self._Join(mdocs, ",", " "), "}")
+
+    def _BashArray(self, varray):
+        # type: (value.BashArray) -> MeasuredDoc
+        if len(varray.strs) == 0:
+            return _Text("(BashArray)")
+        mdocs = []
+        for s in varray.strs:
+            if s is None:
+                mdocs.append(_Text("null"))
+            else:
+                mdocs.append(self._StringLiteral(s))
+        return self._Surrounded("(BashArray", " ", self._Join(mdocs, "", " "),
+                                ")")
+
+    def _BashAssoc(self, vassoc):
+        # type: (value.BashAssoc) -> MeasuredDoc
+        if len(vassoc.d) == 0:
+            return _Text("(BashAssoc)")
+        mdocs = []
+        for k2, v2 in iteritems(vassoc.d):
+            mdocs.append(
+                _Concat([
+                    _Text("["),
+                    self._StringLiteral(k2),
```

Comment:
Note that I'm printing this like a plain string. That effects both the style (not colored the same as a Ysh Dictionary key) and the quotes (quotes are always present, unlike Ysh Dictionary keys which omit quotes if they're identifier-like). It could instead be printed as a _DictKey, which would change both of those properties.